File: ffi/ZMDB.swift
// Swift wrapper for ZMDB
// Usage: import this file in your iOS project

import Foundation

public enum ZMDBError: Error {
    case notFound
    case corruption
    case invalidKey
    case outOfMemory
    case ioError
    case unknown(Int32)
}

public class ZMDBConfig {
    public var cacheSize: Int = 4 * 1024 * 1024
    public var compressionEnabled: Bool = true
    public var syncMode: Int32 = 1 // 0=none, 1=normal, 2=full
    
    public init() {}
}

public class ZMDBDatabase {
    private var handle: OpaquePointer?
    
    public init(path: String, config: ZMDBConfig = ZMDBConfig()) throws {
        var cConfig = zmdb_config_t(
            cache_size: config.cacheSize,
            compression_enabled: config.compressionEnabled ? 1 : 0,
            sync_mode: config.syncMode
        )
        
        var db: OpaquePointer?
        let result = path.withCString { pathPtr in
            zmdb_open(pathPtr, &cConfig, &db)
        }
        
        guard result == ZMDB_OK else {
            throw ZMDBError.from(result)
        }
        
        self.handle = db
    }
    
    deinit {
        if let handle = handle {
            zmdb_close(handle)
        }
    }
    
    public func put(key: Data, value: Data) throws {
        guard let handle = handle else { throw ZMDBError.unknown(-1) }
        
        let result = key.withUnsafeBytes { keyPtr in
            value.withUnsafeBytes { valuePtr in
                zmdb_put(handle, 
                        keyPtr.baseAddress?.assumingMemoryBound(to: UInt8.self),
                        key.count,
                        valuePtr.baseAddress?.assumingMemoryBound(to: UInt8.self),
                        value.count)
            }
        }
        
        guard result == ZMDB_OK else {
            throw ZMDBError.from(result)
        }
    }
    
    public func get(key: Data) throws -> Data {
        guard let handle = handle else { throw ZMDBError.unknown(-1) }
        
        var valuePtr: UnsafeMutablePointer<UInt8>?
        var valueLen: Int = 0
        
        let result = key.withUnsafeBytes { keyPtr in
            zmdb_get(handle,
                    keyPtr.baseAddress?.assumingMemoryBound(to: UInt8.self),
                    key.count,
                    &valuePtr,
                    &valueLen)
        }
        
        guard result == ZMDB_OK else {
            throw ZMDBError.from(result)
        }
        
        guard let ptr = valuePtr else {
            throw ZMDBError.notFound
        }
        
        let data = Data(bytes: ptr, count: valueLen)
        zmdb_free(ptr)
        
        return data
    }
    
    public func delete(key: Data) throws {
        guard let handle = handle else { throw ZMDBError.unknown(-1) }
        
        let result = key.withUnsafeBytes { keyPtr in
            zmdb_delete(handle,
                       keyPtr.baseAddress?.assumingMemoryBound(to: UInt8.self),
                       key.count)
        }
        
        guard result == ZMDB_OK else {
            throw ZMDBError.from(result)
        }
    }
    
    public func contains(key: Data) -> Bool {
        guard let handle = handle else { return false }
        
        return key.withUnsafeBytes { keyPtr in
            zmdb_contains(handle,
                         keyPtr.baseAddress?.assumingMemoryBound(to: UInt8.self),
                         key.count) != 0
        }
    }
    
    public var count: Int {
        guard let handle = handle else { return 0 }
        return zmdb_count(handle)
    }
}

extension ZMDBError {
    static func from(_ code: zmdb_error_t) -> ZMDBError {
        switch code {
        case ZMDB_ERROR_NOT_FOUND: return .notFound
        case ZMDB_ERROR_CORRUPTION: return .corruption
        case ZMDB_ERROR_INVALID_KEY: return .invalidKey
        case ZMDB_ERROR_OUT_OF_MEMORY: return .outOfMemory
        case ZMDB_ERROR_IO: return .ioError
        default: return .unknown(code)
        }
    }
}

File: ffi/ZMDB.java
package com.zmdb;

/**
 * ZMDB - Zig Mobile Database for Android
 * High-performance embedded database optimized for mobile devices
 */
public class ZMDB {
    
    static {
        System.loadLibrary("zmdb");
    }
    
    private long nativeHandle;
    
    public static class Config {
        public int cacheSize = 4 * 1024 * 1024; // 4MB
        public boolean compressionEnabled = true;
        public int syncMode = 1; // 0=none, 1=normal, 2=full
    }
    
    public enum Error {
        OK(0),
        NOT_FOUND(1),
        CORRUPTION(2),
        INVALID_KEY(3),
        OUT_OF_MEMORY(4),
        IO_ERROR(5);
        
        private final int code;
        
        Error(int code) {
            this.code = code;
        }
        
        public static Error fromCode(int code) {
            for (Error e : values()) {
                if (e.code == code) return e;
            }
            return IO_ERROR;
        }
    }
    
    public static class ZMDBException extends Exception {
        private final Error error;
        
        public ZMDBException(Error error) {
            super("ZMDB Error: " + error.name());
            this.error = error;
        }
        
        public Error getError() {
            return error;
        }
    }
    
    /**
     * Open a database
     */
    public ZMDB(String path, Config config) throws ZMDBException {
        int result = nativeOpen(path, config.cacheSize, 
                               config.compressionEnabled ? 1 : 0,
                               config.syncMode);
        if (result != 0) {
            throw new ZMDBException(Error.fromCode(result));
        }
    }
    
    /**
     * Open a database with default configuration
     */
    public ZMDB(String path) throws ZMDBException {
        this(path, new Config());
    }
    
    /**
     * Close the database
     */
    public void close() {
        if (nativeHandle != 0) {
            nativeClose(nativeHandle);
            nativeHandle = 0;
        }
    }
    
    /**
     * Store a key-value pair
     */
    public void put(byte[] key, byte[] value) throws ZMDBException {
        int result = nativePut(nativeHandle, key, value);
        if (result != 0) {
            throw new ZMDBException(Error.fromCode(result));
        }
    }
    
    /**
     * Retrieve a value by key
     */
    public byte[] get(byte[] key) throws ZMDBException {
        byte[] value = nativeGet(nativeHandle, key);
        if (value == null) {
            throw new ZMDBException(Error.NOT_FOUND);
        }
        return value;
    }
    
    /**
     * Delete a key-value pair
     */
    public void delete(byte[] key) throws ZMDBException {
        int result = nativeDelete(nativeHandle, key);
        if (result != 0) {
            throw new ZMDBException(Error.fromCode(result));
        }
    }
    
    /**
     * Check if a key exists
     */
    public boolean contains(byte[] key) {
        return nativeContains(nativeHandle, key);
    }
    
    /**
     * Get the number of entries
     */
    public long count() {
        return nativeCount(nativeHandle);
    }
    
    @Override
    protected void finalize() throws Throwable {
        close();
        super.finalize();
    }
    
    // Native methods
    private native int nativeOpen(String path, int cacheSize, int compression, int syncMode);
    private native void nativeClose(long handle);
    private native int nativePut(long handle, byte[] key, byte[] value);
    private native byte[] nativeGet(long handle, byte[] key);
    private native int nativeDelete(long handle, byte[] key);
    private native boolean nativeContains(long handle, byte[] key);
    private native long nativeCount(long handle);
}
